### PR TITLE
Improve relay selection to prioritize active reservations

### PR DESCRIPTION
**Pull request by parth-soni07**
## What was wrong?

Issue #735

## How was it fixed?
Prioritizes first available relay with reservations instead of returning obvious first one.

### CODE REVIEW

**Comment by paschal533** [comment]
Hi @parth-soni07  Thanks for tackling this issue. I really appreciate you diving into the relay selection logic. Overall this is a solid first pass and I like the direction you're taking. The test coverage is especially impressive 👏

I do have a few concerns I'd like you to address

The `relay_counter` isn't protected against concurrent access. Since `_select_relay()` can be called from multiple places simultaneously, we could have race conditions. Let's wrap it with a lock:

```python
self._relay_counter = 0
self._relay_counter_lock = trio.Lock()
```

Then use it like:
```python
async with self._relay_counter_lock:
    self._relay_counter += 1
    index = (self._relay_counter - 1) % len(relays_with_reservations)
```

Right now the counter increments every time we call `_select_relay()`, even if we don't find any relays. This means the retry loop will keep bumping the counter, leading to uneven distribution. Could you move the increment to only happen when we successfully return a relay?

Calling `get_relay_info()` in a loop for every selection might get expensive if we have lots of relays. Not a blocker for now, but worth thinking about, maybe we could maintain a cached list of relays with reservations? Just food for thought.

Could you mention the round-robin aspect more explicitly in the newsfragment? Something like:

```rst
Implemented round-robin load balancing for CircuitV2 relay selection, prioritizing relays with active reservations for more reliable and evenly distributed relay usage.
```

Again, really solid work here, the test suite especially shows you're thinking through the edge cases carefully. Once we address the thread safety and counter issues, this should be good to go. Well done @parth-soni07 keep it up with the good work

**Comment by paschal533** [comment]
Hi @parth-soni07, Thanks for working on this, I can see you are putting good effort into it.
I do have some feedback though. While I appreciate the round-robin approach as an improvement over "first available", I think we should actually aim higher here since we're already making changes.

Round-robin distributes load evenly, but it doesn't actually help users get the best relay for their connection. A relay could be geographically far away, overloaded, or having network issues, and round-robin would still send traffic to it equally.
Instead, I'd like you to consider implementing a performance-based selection strategy using either:

1. Least response time - Track how long it takes to establish connections through each relay
2. Least connections - Track active circuit counts per relay
3. Hybrid approach - Combine both metrics with reservation priority

This would give us better user experience, natural load balancing, self-healing, and still prioritizes reserved relays when performance is similar

I'm happy to pair with you on this if you need any help

CCing @seetadev, @pacrob and @acul71

**Comment by parth-soni07** [comment]
> Hi @parth-soni07, Thanks for working on this, I can see you are putting good effort into it. I do have some feedback though. While I appreciate the round-robin approach as an improvement over "first available", I think we should actually aim higher here since we're already making changes.
> 
> Round-robin distributes load evenly, but it doesn't actually help users get the best relay for their connection. A relay could be geographically far away, overloaded, or having network issues, and round-robin would still send traffic to it equally. Instead, I'd like you to consider implementing a performance-based selection strategy using either:
> 
> 1. Least response time - Track how long it takes to establish connections through each relay
> 2. Least connections - Track active circuit counts per relay
> 3. Hybrid approach - Combine both metrics with reservation priority
> 
> This would give us better user experience, natural load balancing, self-healing, and still prioritizes reserved relays when performance is similar
> 
> I'm happy to pair with you on this if you need any help
> 
> CCing @seetadev, @pacrob and @acul71

Hey @paschal533 that is indeed a great idea to explore a more precise way that target the users in an effective manner. We can surely work on this to come out with a better solution. Thank You!

**Comment by parth-soni07** [comment]
based on @paschal533 's last feedback from the discussion: https://github.com/libp2p/py-libp2p/discussions/1043, I have made the requested changes and would request another feedback from @paschal533 @seetadev 
 
**Core Implementation**: ✅ Complete (15/15)
**Testing**: ✅ Complete (9/9) - 26 tests passing
**Integration Tests**: ✅ All existing relay selection tests passing (3/3)
**Enhancements**: ⚠️ Optional

---

## Implementation Summary

### RelayPerformanceTracker Class

**Location**: `libp2p/relay/circuit_v2/performance_tracker.py`

**Key Features**:
- Per-relay performance tracking (latency EMA, success rate, active circuits)
- Thread-safe operations with locks
- Configurable scoring parameters
- Round-robin fallback for equal scores

**Methods Implemented**:
1. ✅ `record_connection_attempt(relay_id, latency_ms, success)` - Tracks latency with EMA and success/failure
2. ✅ `record_circuit_opened(relay_id)` - Increments active circuit count
3. ✅ `record_circuit_closed(relay_id)` - Decrements active circuit count
4. ✅ `get_relay_score(relay_id)` - Returns score based on latency, load, and success rate (lower is better)
5. ✅ `select_best_relay(available_relays)` - Selects best relay with reservation preference support

**Scoring Formula**:
- Base score = latency_ema_ms × latency_penalty_ms
- + circuit_penalty = active_circuits × circuit_penalty_multiplier
- + failure_penalty = failure_count × failure_penalty
- - success_bonus = success_rate × success_bonus_multiplier
- Returns `float('inf')` for unhealthy/overloaded relays

**Integration**:
- ✅ Integrated into `CircuitV2Transport`
- ✅ Connection attempts tracked in `dial_peer_info()`
- ✅ Circuit opened events tracked
- ✅ `_select_relay()` uses tracker with round-robin fallback for equal scores

Screenshot 2025-12-01 at 1 37 42 PM

**Comment by paschal533** [comment]
Hi @parth-soni07  

Excellent work this is exactly what I was looking for. The `RelayPerformanceTracker` implementation looks solid and you've nailed the architecture I had in mind. Really appreciate how you pivoted from the Prometheus-heavy approach to this clean, focused solution.

I have a  few questions/suggestions:

1. **EMA initialization:** I see you're handling the first measurement specially (`if stats.latency_ema_ms == 0.0`), but what if a relay legitimately has 0ms latency (localhost testing)? Might be worth using a sentinel value like `-1.0` or tracking whether we've seen the first measurement separately. Not a blocker, just something to consider.

2. **Circuit tracking:** How are you handling circuit closures that happen automatically? I see you have `record_circuit_closed()`, but are you calling this everywhere circuits are cleaned up (timeouts, errors, normal completion)? Might be worth a quick audit to make sure we're not leaking circuit counts.

3. **Export metrics method:** I see you have `export_metrics()` returning a dict - nice forward-thinking for when we do add Prometheus later. Are you planning to wire this up to anything right now, or is it just there for future use?

If you can address these questions (most are just clarifications, not blockers), I think we're ready to merge this. After merge, let's plan to:
1. Monitor the relay selection behavior in a test environment
2. Tune the scoring parameters based on real data
3. Then circle back to add Prometheus integration as a separate PR

Really solid work @parth-soni07, you took the feedback well and delivered exactly what we needed. The architecture is clean and extensible. Let me know your thoughts on the questions above and we should be good to go! 👏

cc @seetadev for final review

**Comment by parth-soni07** [comment]
Hey @paschal533, thanks for the great feedback! I've addressed all three points:

## 1. EMA Initialization - ✅ Fixed

**Issue**: Using `0.0` as a sentinel value would incorrectly treat legitimate 0ms latency (e.g., localhost) as "no data yet".

**Solution Implemented**: 
- Changed default value from `0.0` to `-1.0` as a sentinel value
- Updated first measurement check from `if stats.latency_ema_ms == 0.0:` to `if stats.latency_ema_ms < 0.0:`
- Added handling in `get_relay_score()` to return `unknown_relay_score` when `latency_ema_ms < 0.0` (no latency data yet)

**Changes**:
- `RelayStats.latency_ema_ms` now defaults to `-1.0` with comment: `# Exponential moving average latency (-1.0 = no data yet)`
- Added test `test_record_connection_attempt_zero_latency()` to verify 0ms latency is handled correctly
- Added test `test_get_relay_score_no_latency_data()` to verify sentinel value behavior

**Result**: Legitimate 0ms latency is now properly tracked and won't be confused with "no data yet".

---

## 2. Circuit Closure Tracking - ✅ Fixed

**Issue**: `record_circuit_closed()` existed but was never called, causing circuit counts to grow unbounded and eventually making all relays appear overloaded.

**Solution Implemented**: 
Created a `TrackedRawConnection` wrapper class that automatically tracks circuit closure:

- **Wrapper class**: `TrackedRawConnection` wraps `RawConnection` and intercepts `close()` calls
- **Automatic tracking**: When `close()` is called, it automatically calls `record_circuit_closed()` before delegating to the wrapped connection
- **Double-close protection**: Uses `_closed` flag to ensure circuit is only decremented once, even if `close()` is called multiple times
- **Transparent delegation**: All other methods/attributes delegate to the wrapped connection via `__getattr__`

**Changes**:
- Added `TrackedRawConnection` class in `libp2p/relay/circuit_v2/transport.py`
- Updated `dial_peer_info()` to return `TrackedRawConnection` instead of `RawConnection`
- Changed return type to `IRawConnection` for type compatibility
- Added 3 comprehensive unit tests covering:
  - Normal closure tracking
  - Double-close protection
  - Method delegation

**Result**: Circuit closures are now automatically tracked for all connections created through `dial_peer_info()`, preventing unbounded growth of circuit counts. The wrapper handles all closure paths (normal, error, timeout) since it intercepts the `close()` method.

**Note**: This solution is non-intrusive - it doesn't modify core `RawConnection` class and is transparent to callers (returns `IRawConnection` interface).

---

## 3. Export Metrics Method - ✅ Working as Intended

**Status**: The `export_metrics()` method is intentionally not wired up to anything currently. It's designed for future Prometheus integration.

**Current Implementation**:
- Returns a dictionary with relay statistics (latency, success rates, active circuits, etc.)
- Ready to be called when Prometheus integration is added in a future phase
- Not currently used by any monitoring system

**No changes needed** - this is working as designed per the original feedback to defer Prometheus integration.

---

## Summary

Both critical issues (EMA initialization and circuit closure tracking) have been fixed with comprehensive test coverage. All tests pass (31 total: 28 performance tracker tests + 3 new wrapper tests), and all lint/type checks pass.

The implementation is ready and addresses the concerns raised. Ready for review!

**Comment by yashksaini-coder** [comment]
@parth-soni07 @paschal533 @seetadev

# PR #972 Review Summary — Smarter Relay Selection for CircuitV2

This PR delivers a much-needed upgrade to how CircuitV2 selects relays. Previously, the transport simply picked the first available relay — no balancing, no performance checks, and no awareness of active reservations.  
This change fixes all of that.

**TL;DR:**  
➡️ Relay selection is now smarter, fairer, more reliable, and reservation-aware.  
➡️ The implementation is complete and well-tested.  
➡️ Everything requested in Issue #735 is addressed.

---

## What This PR Achieves

| Goal | Status | Notes |
|------|:------:|-------|
| Avoid overloading the first relay | ✅ | Added load balancing + active circuit tracking |
| Prefer faster/more reliable relays | ✅ | Latency EMA + scoring |
| Track relay health | ✅ | Success/failure stats |
| Use long-term performance history | ✅ | New RelayPerformanceTracker |
| Prioritize relays with reservations | ✅ | First-class support |
| Ensure thread-safe counter updates | ✅ | Lock-guarded |
| Track circuit closures | ✅ | TrackedRawConnection wrapper |

All items from Issue #735 are fully implemented.

---

## Key Additions

### 🔧 `RelayPerformanceTracker`

A new module that keeps per-relay stats: latency, success rate, and active circuits.  
It also computes a score for each relay so the selector can pick the best one.

The scoring formula blends:

- latency (EMA)  
- active circuit count  
- failure history  
- success rate bonus  

Unhealthy/overloaded relays are automatically deprioritized.

---

### 🎛️ Smarter Relay Selection

Relay selection now runs through a clear pipeline:

1. Collect all discovered relays  
2. Split them into **with reservation** and **without**  
3. Score each relay using the performance tracker  
4. Pick the best one  
5. If scores tie → use round-robin (thread-safe)  

This is a major improvement over the "just pick the first one" behavior.

---

### 🔌 `TrackedRawConnection`

Connections are wrapped so that when they’re closed, the tracker decrements the active circuit count.  
Prevents relay load metrics from drifting over time.

---

## Test Coverage

This PR brings **extensive** test coverage:

- ~28 tests for `RelayPerformanceTracker`  
- Connection wrapper tests (close, double-close, delegation)  
- Relay discovery tests, including reservation behavior  
- Integration tests verifying end-to-end relay selection, fallback logic, and round-robin  

Everything passes under Linux; Windows CI is still in progress.

---

## Documentation & Changelog

Docs updated for CircuitV2.  
Changelog entry added summarizing the new behavior:

> “Added round-robin relay selection with prioritization of relays that  
> hold active reservations.”

---

## Reviewer Feedback Resolved

All earlier review notes (thread safety, EMA sentinel, counter behavior, etc.) were addressed cleanly.

---

## Overall Verdict

This PR brings a **major quality upgrade** to CircuitV2.  
Relay selection is now balanced, reservation-aware, statistically informed, and tracked over time.  
The code is well-structured and the tests are thorough.

**Recommendation:** ✔️ **Ready to merge** The Failinf CI check is of QUIC transport test, which we have not committed or work in this PR.

**Comment by yashksaini-coder** [comment]
@parth-soni07 Hi could you please review these merge conflicts ?

**Comment by parth-soni07** [comment]
> @parth-soni07 Hi could you please review these merge conflicts ?

hey @yashksaini-coder , yes I'm currently working on the conflicts as they are functional changes and this took me slightly more than usual to fix, keeping my current progress on this in mind, I think I'll push the changes in a few, more hours. Thanks for the gentle reminder though :)

**Comment by yashksaini-coder** [comment]
> > @parth-soni07 Hi could you please review these merge conflicts ?
> 
> hey @yashksaini-coder , yes I'm currently working on the conflicts as they are functional changes and this took me slightly more than usual to fix, keeping my current progress on this in mind, I think I'll push the changes in a few, more hours. Thanks for the gentle reminder though :)

Great I'm in no hurry, we just make sure to handle this and manage it, if any fail ci arrives I will look into those and share my findings.

**Comment by parth-soni07** [comment]
# Changelog: Merge Conflict Resolution

## Summary
Resolved merge conflicts between `feat/improve-relay-selection` and `main` branches, integrating performance tracking with stored addresses and DHT discovery features.

## Changes

### `libp2p/relay/circuit_v2/transport.py`
- **Integrated performance tracking with stored addresses**: Stored circuit addresses are tried first (fast path), with performance metrics tracked for all attempts
- **Added `TrackedRawConnection` wrapper**: Automatically tracks circuit lifecycle to maintain accurate active circuit counts
- **Hybrid relay selection**: Combines performance-based selection with DHT discovery fallback
- **Enhanced `dial_peer_info`**: Now tracks latency and success/failure for all connection attempts, including stored address attempts
- **Updated `_select_relay`**: Uses `RelayPerformanceTracker` for intelligent relay selection with round-robin for equal scores

### `tests/core/relay/test_circuit_v2_transport.py`
- **Updated tests for `TrackedRawConnection`**: Tests now expect `TrackedRawConnection` instead of `RawConnection`
- **Migrated to performance tracker API**: Replaced legacy metrics (`_relay_metrics`, `_relay_list`) with `performance_tracker` methods
- **Updated test assertions**: All selection tests now use `record_connection_attempt()` and `get_relay_stats()` for verification

## Technical Details
- Performance tracking records latency, success rate, and active circuit counts per relay
- Stored addresses provide fast-path optimization for previously connected peers
- DHT discovery acts as fallback when no relays are available from discovery
- Round-robin selection applied for relays with equal performance scores

The conflicts are now resolved @yashksaini-coder @seetadev @paschal533. Thanks :)

**Comment by seetadev** [comment]
@parth-soni07 : Wonderful, thank you Parth for resolving the merge conflicts. Appreciate the efforts.

Doing a detailed review today evening.

Wish to also invite @pacrob , @acul71, @sumanjeet0012 and @Winter-Soren to share their reviews and feedback.

It indeed impacts PRs by @Winter-Soren and @sumanjeet0012. Please do share feedback before we do a final review + merge.

File: libp2p/relay/circuit_v2/transport.py [hunk 1]
```diff
@@ -315,13 +418,45 @@ async def dial_peer_info(
...
 
             if status_code != StatusCode.OK:
+                # Record failed connection attempt
+                latency_ms = (trio.current_time() - connection_start_time) * 1000
+                self.performance_tracker.record_connection_attempt(
+                    relay_id=relay_peer_id,
+                    latency_ms=latency_ms,
+                    success=False,
+                )
                 raise ConnectionError(f"Relay connection failed: {status_msg}")
```
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py]: It looks like this will record the failed connection attempt, then the `raise` will drop it to the exception on line 452, which then records another failed attempt. This seems like it will inflate the failure count and mark a relay unhealthy too soon.

File: libp2p/relay/circuit_v2/transport.py [hunk 2]
```diff
@@ -246,19 +321,47 @@ async def dial_peer_info(
                 conn = await self._dial_via_circuit_addr(ma, dest_info)
                 if conn:
                     logger.debug("Connected via stored circuit addr %s", ma)
+                    # Record successful connection attempt
+                    relay_peer_id = self._extract_relay_id_from_ma(ma)
+                    latency_ms = (trio.current_time() - connection_start_time) * 1000
```
Comment (pacrob) [inline libp2p/relay/circuit_v2/transport.py]: We start the timing with `connection_start_time`, but then step through `circuit_addrs`, meaning `ma`s that appear later in the list will look like they have longer latency than they actually do. Would it make sense to have a start time for each connection attempt? Or am I understanding it incorrectly?

**Comment by parth-soni07** [comment]
Thanks @pacrob for your review. Both points were spot on. I've fixed both issues:

1. Removed the duplicate failure recording - the exception handler now handles all failures, preventing double-counting.
2. Fixed the latency timing - each connection attempt in the loop now uses its own `attempt_start_time` measured right before the attempt, so latency is accurate per relay rather than cumulative.

The changes are in the latest commit. Let me know if you'd like me to adjust anything else!

**Review by pacrob — APPROVED** [review]
lgtm!

**Comment by seetadev** [comment]
@parth-soni07 : Thanks for the update — and great work addressing the feedback. 🙌

@pacrob : Thank you for the thoughtful review and for approving the PR. 

In addition to the fixes around failure accounting and per-attempt latency timing, the update to prioritize the first available relay with existing reservations is an important improvement over returning the obvious first candidate.

Wish to also thank @yashksaini-coder and @paschal533 for their important contributions to this PR. Appreciate your efforts.
